### PR TITLE
Bug 706809 - write through cache for the sauropod backend (draft)

### PR DESCRIPTION
**Pull request by tarekziade**
We maintain the meta document in memcached to avoid hitting Sauropod on most GETs

### CODE REVIEW

File: appsync/storage/sauropod.py [hunk 1]
```diff
@@ -132,13 +146,50 @@ def _resume_session(self, token):
...
+        session.set(key, json.dumps(data), if_match=etag)
+
+        if self.cache:
+            doc = session.getitem(key)
+            cache_key = _key(user, collection, 'meta')
+            self.cache.set(cache_key, doc, self.cache_ttl)
+
+    def _get_cached_metadata(self, session, user, collection):
+        # getting the cached value if possible
+        if self.cache is None:
```
Comment (rfk) [inline appsync/storage/sauropod.py]: Should be "is not None"?

File: appsync/storage/sauropod.py [hunk 2]
```diff
@@ -132,13 +146,50 @@ def _resume_session(self, token):
...
+            return
+        cache_key = _key(user, collection, 'meta')
+        self.cache.delete(cache_key)
+
+    def _set_cached_metadata(self, session, user, collection, data, etag):
+        key = collection + "::meta"
+        session.set(key, json.dumps(data), if_match=etag)
+
+        if self.cache:
+            doc = session.getitem(key)
```
Comment (rfk) [inline appsync/storage/sauropod.py]: the call to session.set() above should return the updated item, so you can avoid fetching it against here

**Comment by rfk** [comment]
@dclarke: the _get_cached_metadata method calls through to hbase itself, so I guess it should just grow some logic to ignore e.g. connection errors from the cache.

File: appsync/storage/sauropod.py [hunk 3]
```diff
@@ -132,13 +146,50 @@ def _resume_session(self, token):
...
+                return cached
+
+        # getting the meta document.
+        doc = session.getitem(collection + "::meta")
+
+        # set the cache
+        if self.cache is None:
+            self.cache.set(cache_key, doc, time=self.cache_ttl)
+
+        return res
```
Comment (rfk) [inline appsync/storage/sauropod.py]: res undefined, should be "return doc"?

**Comment by tarekziade** [comment]
@dclarke this code has already been merged......

Not this is not merged, it's a ongoing work, a draft in my clone

**Comment by dclarke** [comment]
rfk, looks correct

File: appsync/storage/sauropod.py (hunk 2 above)
Comment (tarekziade) [inline appsync/storage/sauropod.py]: Oh, I just realized set() returns the Item. So yeah, changing this

**Comment by dclarke** [comment]
tarek: yup correct , i read incorrectly, looks good though

**Comment by tarekziade** [comment]
@dclarke Adding a hit counter test now

**Comment by tarekziade** [comment]
@dclarke in this last commit, I am counting the hits. 
- If Memcache is running the tests will use it and make sure it caches Sauropod GETs on the meta document, and saves hits.
- If Memcache is not running, the tests makes sure the backend still works with no cache.
- For the SQL backend, Memcache is mandatory for the tokens, if not running a verify will generate a 503 and another call a 401.

I think the change is ready to be commited now, let me know guys.

File: appsync/storage/sauropod.py [hunk 4]
```diff
@@ -132,13 +148,63 @@ def _resume_session(self, token):
...
+    def _set_cached_metadata(self, session, user, collection, data, etag):
+        key = collection + "::meta"
+        doc = session.set(key, json.dumps(data), if_match=etag)
+
+        if self.cache is not None:
+            cache_key = _key(user, collection, 'meta')
+            try:
+                self.cache.set(cache_key, doc, self.cache_ttl)
+            except CacheError:
+                logger.error('Unable to read the metadata in the cache.')
```
Comment (rfk) [inline appsync/storage/sauropod.py]: s/read/write/

**Pull request merged**